wgsl/index.bs

```diff
@@ -454,6 +454,167 @@ whether failure to meet a particular requirement
+
+A <dfn dfn-for="diagnostic">triggering rule</dfn> is condition on a source program that, when met, causes a [=diagnostic=] to be generated.
+The following table defines the standard triggering rules.
+
+<table class='data'>
+  <caption>Diagnostic triggering rules defined in WGSL</caption>
+  <thead>
+    <tr><th>Triggering rule<th>Default Severity<th>Triggering Location<th>Description
+  </thead>
+
+  <tr>
+    <td><dfn noexport dfn-for="trigger">derivative_uniformity</dfn>
+    <td>[=severity/error=]
+    <td>A call to a [=builtin functions that compute a derivative|builtin function that computes a derivative=].
+        That is, a call to any of:
+        * the [[#derivative-builtin-functions|derivative builtin functions]], or
+        * [[#texturesample|textureSample]], or
+        * [[#texturesamplebias|textureSampleBias]], or
+        * [[#texturesamplecompare|textureSampleCompare]].
+
+    <td>A call to a builtin function computes derivatives, but [=uniformity analysis=] cannot prove that the call occurs in [=uniform control flow=].
+
+    See [[#uniformity]]
+</table>
+
+An implementation may support triggering rules not specified here, provided they are spelled differently from the standard triggering rules.
+A [=diagnostic filter=] with an unrecognized [=diagnostic/triggering rule=] [=behavioral requirement|will=] be ignored.
+
+Future versions of this specification may remove a particular rule or weaken its default severity
+(i.e. replace its current default with a less severe default) and still be deemed as satisfying backward compatibility.
+For example, a future version of WGSL may change the default severity for [=trigger/derivative_uniformity=] from `error` to either `warning` or `info`.
+After such a change to the specification, previously valid programs would remain valid.
+
+### Diagnostic Filtering ### {#diagnostic-filtering}
+
+Once a [=diagnostic=] is [=triggered=], WGSL provides filtering mechanisms to discard the diagnostic, or to modify its severity.
+
+A <dfn noexport>diagnostic filter</dfn> *DF* has three parameters:
+* *AR*: a range of source text known as the <dfn noexport>affected range</dfn>
+* *NS*: a new severity
+* *TR*: a [=diagnostic/triggering rule=]
+
+Applying a diagnostic filter *DF*(*AR*,*NS*,*TR*) to a diagnostic *D* has the following effect:
+* If *D*'s [=diagnostic/triggering location=] is in *AR* and *D*'s [=diagnostic/triggering rule=] is *TR*, then:
+    * If *NS* is `off`, then *D* is discarded (i.e. deleted, or forgotten).
+    * Otherwise, *NS* is either `error` or `warning`.
+        Set *D*'s [=diagnostic/severity=] property to *NS*.
+* Otherwise, *D* is unchanged.
+
+A <dfn noexport>range diagnostic filter</dfn> is a [=diagnostic filter=] that is applied to each
+diagnostic whose [=diagnostic/triggering location=] in a specified range of source text.
+A range diagnostic filter is specified as a `@diagnostic` [=attribute=] immediately before the affected source range:
```

Comment:
Can filters be applied to arbitrary statements? Usually the "Attributes" section restricts where the attribute can appear, but the description of this attribute doesn't seem to have any rules. It's true that the immediately following text talks about call phrases, but not in a way that seems to exclude placement elsewhere.

It would make sense to me to put them on statements and functions, too.

Reply:
Since it's expressed as an attribute, the grammar tells us where the diagnostic can appear.

In this PR I went with the minimal thing needed for uniformity-opt-out:  attributes can now be put on a 'call_expr' which is a grammar element introduced in this PR.  (Yes, it's subtle and I didn't call it out.)

Internally we very much agreed it would be nice to allow the general mechanism on statements in general, particularly compound statements, and function declarations.  Certainly that's within the design space.  
The spec would also need to describe the affected range for each of those cases. (They're pretty obvious.)

I suggest we do that later.  It also leads to extra CTS testing.

Reply:
> It would make sense to me to put them on statements and functions, too. 

Let's file that as a post-v1 enhancement.

Reply:
Filed.  https://github.com/gpuweb/gpuweb/issues/3689 
Per offline discussion today, it's set for v1